packages/@lwc/engine-core/src/framework/hydration.ts

```diff
@@ -395,10 +396,28 @@ function validateAttrs(vnode: VBaseElement, elm: Element, renderer: RendererAPI)
 }
 
 function validateClassAttr(vnode: VBaseElement, elm: Element, renderer: RendererAPI): boolean {
-    const {
-        data: { className, classMap },
-    } = vnode;
+    const { data, owner } = vnode;
+    let { className, classMap } = data;
     const { getProperty, getClassList } = renderer;
+    const scopedToken = getScopeTokenClass(owner);
+
+    // Classnames for scoped CSS are added directly to the DOM during rendering,
+    // or to the VDOM on the server in the case of SSR. As such, these classnames
+    // are never present in VDOM nodes in the browser.
+    //
+    // Consequently, hydration mismatches will occur if scoped CSS token classnames
+    // are rendered during SSR. This needs to be accounted for when validating.
+    if (scopedToken) {
```

Comment:
Typically we would do `if (!isUndefined(scopedToken))` but we are super inconsistent about it. I believe it has some theoretical perf benefit, at least in microbenchmarks.